components/fxa-client/src/internal/auth_state.rs

```diff
@@ -0,0 +1,213 @@
+            .remove(commands::send_tab::COMMAND_NAME);
+    }
+
+    pub fn last_handled_command_index(&self) -> Option<u64> {
+        self.persisted_state.last_handled_command
+    }
+
+    pub fn set_last_handled_command_index(&mut self, idx: u64) {
+        self.persisted_state.last_handled_command = Some(idx)
+    }
+
+    pub fn current_device_id(&self) -> Option<&str> {
+        self.persisted_state.current_device_id.as_deref()
+    }
+
+    pub fn set_current_device_id(&mut self, device_id: String) {
+        self.persisted_state.current_device_id = Some(device_id);
+    }
+
+    pub fn get_scoped_key(&self, scope: &str) -> Option<&ScopedKey> {
+        self.persisted_state.scoped_keys.get(scope)
+    }
+
+    pub(crate) fn last_seen_profile(&self) -> Option<&CachedResponse<Profile>> {
+        self.persisted_state.last_seen_profile.as_ref()
+    }
+
+    pub(crate) fn set_last_seen_profile(&mut self, profile: CachedResponse<Profile>) {
+        self.persisted_state.last_seen_profile = Some(profile)
+    }
+
+    pub fn clear_last_seen_profile(&mut self) {
+        self.persisted_state.last_seen_profile = None
+    }
+
+    pub fn get_cached_access_token(&mut self, scope: &str) -> Option<&AccessTokenInfo> {
+        self.persisted_state.access_token_cache.get(scope)
+    }
+
+    pub fn add_cached_access_token(&mut self, scope: impl Into<String>, token: AccessTokenInfo) {
+        self.persisted_state
+            .access_token_cache
+            .insert(scope.into(), token);
+    }
+
+    pub fn clear_access_token_cache(&mut self) {
+        self.persisted_state.access_token_cache.clear()
+    }
+
+    /// Begin an OAuth flow.  This saves the OAuthFlow for later.  `state` must be unique to this
+    /// oauth flow process.
+    pub fn begin_oauth_flow(&mut self, state: impl Into<String>, flow: OAuthFlow) {
+        self.flow_store.insert(state.into(), flow);
+    }
+
+    /// Get an OAuthFlow from a previous `begin_oauth_flow()` call
+    ///
+    /// This operation removes the OAuthFlow from the our internal map.  It can only be called once
+    /// per `state` value.
+    pub fn get_oauth_flow(&mut self, state: &str) -> Option<OAuthFlow> {
+        self.flow_store.remove(state)
+    }
+
+    /// Complete an OAuth flow.
+    pub fn complete_oauth_flow(
+        &mut self,
+        scoped_keys: Vec<(String, ScopedKey)>,
+        refresh_token: RefreshToken,
+        session_token: Option<String>,
+    ) {
+        // When our keys change, we might need to re-register device capabilities with the server.
+        // Ensure that this happens on the next call to ensure_capabilities.
+        self.persisted_state.device_capabilities.clear();
+
+        for (scope, key) in scoped_keys {
+            self.persisted_state.scoped_keys.insert(scope, key);
+        }
+        self.persisted_state.refresh_token = Some(refresh_token);
+        self.persisted_state.session_token = session_token;
+        self.flow_store.clear();
```

Comment:
This is a small functional change.  Once we complete one oauth flow, we delete any other oauth flows that were started.  I don't think it makes sense to allow a user to disconnect, then complete an old oauth flow.